Fixes directory escaping bug in FileServer

despeset opened the pull request:
Closes #37

despeset committed:
Fixes directory escaping bug in FileServer

Closes #37

zachallaun commented on Meddle/src/Meddle.jl:
```diff
@@ -91,14 +91,12 @@ end
         if m != nothing
-            path = normpath(root, m.captures[1])
+            # protect against dir-escaping "/../" commands
+            path = normpath(root, replace(m.captures[1],r"/?\.+/",""))
```
Seems like this may not allow "/some/directory./file"

astrieanna replied:
Rather than modifying the input path, I would check if the `normpath`'d path was in the directory or not. (which is a matter of comparing prefixes).
If it's not in the directory send up a 403 or 404. We don't want to serve some other file if they are _trying_ to escape the directory.

despeset committed:
Improves escape path detection, returns 400

- Now requests like "/abc../abc./abc.md" pass test.
- Any requests that do have "/../" escapes are served a 400: Bad Request
- Abstracted detection to function so it can be unit tested

despeset commented:
```
> mkdir a..
> echo "Hi!" > a../b.cd..e.
> julia examples/meddle.jl
Listening on port 8000...
...
> curl http://localhost:8000/a../b.cd..e.

Hi!
> curl http://localhost:8000/../webstack.jl/a../b.cd..e.

400 Bad Request
```

despeset committed:
Changes escape detection to normalized prefix comp

@astrieanna makes a good case for wanting to preserve legitimate uses of '/../' that do not result in directory escape.  Results in 1 less problem.

despeset commented:
Merged by 8e6bf3881188bfee2abac5cd667e727ef065ac5c
